File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/bridge/solution/RightAngledQuadrilateral.m
% RightAngledQuadrilateral.m

classdef RightAngledQuadrilateral < handle
	properties (Access = private)
		Body % RightAngledQuadrilateralBody
	end

	methods
		% constructor:
		function theRightAngledQuadrilateral = RightAngledQuadrilateral(aHeight, aWidth)
			if nargin == 1
				% aHeight means aSide:
				theRightAngledQuadrilateral.Body = theRightAngledQuadrilateral.makeBody(aHeight, aHeight);
				% or:
				theRightAngledQuadrilateral.Body = SquareBody(aHeight);
			else
				theRightAngledQuadrilateral.Body = theRightAngledQuadrilateral.makeBody(aHeight, aWidth);
			end
		end

		function setHeight(theRightAngledQuadrilateral, aHeight)
			theRightAngledQuadrilateral.Body = theRightAngledQuadrilateral.makeBody(aHeight, theRightAngledQuadrilateral.getWidth());
		end

		function setWidth(theRightAngledQuadrilateral, aWidth)
			theRightAngledQuadrilateral.Body = theRightAngledQuadrilateral.makeBody(theRightAngledQuadrilateral.getHeight(), aWidth);
		end

		function setSide(theRightAngledQuadrilateral, aSide)
			theRightAngledQuadrilateral.Body = theRightAngledQuadrilateral.makeBody(aSide, aSide);
			% or:
			theRightAngledQuadrilateral.Body = SquareBody(aSide);
		end

		function result = getHeight(theRightAngledQuadrilateral)
			result = theRightAngledQuadrilateral.Body.getHeight;
		end

		function result = getWidth(theRightAngledQuadrilateral)
			result = theRightAngledQuadrilateral.Body.getWidth;
		end

		function result = isSquare(theRightAngledQuadrilateral)
			result = theRightAngledQuadrilateral.Body.isSquare;
		end
	end

	methods (Access = private, Static)
		function result = makeBody(aHeight, aWidth)
			if aHeight == aWidth
				result = SquareBody(aHeight);
			else
				result = RectangleBody(aHeight, aWidth);
			end
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/bridge/solution/right_angled_quadrilateral.cpp
// right_angled_quadrilateral.cpp

#include "rectangle_body.h"
#include "right_angled_quadrilateral.h"
#include "right_angled_quadrilateral_body.h"
#include "square_body.h"

right_angled_quadrilateral::right_angled_quadrilateral(double a_height, double a_width)
{
	my_body = make_body(a_height, a_width);
}

right_angled_quadrilateral::right_angled_quadrilateral(double a_side)
{
	my_body = make_body(a_side, a_side);
	// or:
	my_body = new square_body(a_side);
}

void right_angled_quadrilateral::set_height(double a_height)
{
	my_body = make_body(a_height, get_width());
}

void right_angled_quadrilateral::set_width(double a_width)
{
	my_body = make_body(get_height(), a_width);
}

void right_angled_quadrilateral::set_side(double a_side)
{
	my_body = make_body(a_side, a_side);
	// or:
	my_body = new square_body(a_side);
}

double right_angled_quadrilateral::get_height() const
{
	return my_body->get_height();
}

double right_angled_quadrilateral::get_width() const
{
	return my_body->get_width();
}

bool right_angled_quadrilateral::is_square() const
{
	return my_body->is_square();
}

right_angled_quadrilateral_body * right_angled_quadrilateral::make_body(double a_height, double a_width)
{
	if (a_height == a_width)
	{
		return new square_body(a_height);
	}
	else
	{
		return new rectangle_body(a_height, a_width);
	}
}